alert-common/src/main/java/com/synopsys/integration/alert/common/workflow/processor/DigestMessageContentProcessor.java

```diff
@@ -23,91 +23,119 @@
 package com.synopsys.integration.alert.common.workflow.processor;
 
 import java.util.ArrayList;
-import java.util.Collection;
-import java.util.EnumMap;
 import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
-import java.util.Optional;
+import java.util.Set;
+import java.util.SortedSet;
 import java.util.TreeSet;
-import java.util.function.BiFunction;
+import java.util.stream.Collectors;
 
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Component;
 
 import com.synopsys.integration.alert.common.enumeration.FormatType;
-import com.synopsys.integration.alert.common.enumeration.ItemOperation;
 import com.synopsys.integration.alert.common.message.model.AggregateMessageContent;
 import com.synopsys.integration.alert.common.message.model.CategoryItem;
 import com.synopsys.integration.alert.common.message.model.CategoryKey;
 import com.synopsys.integration.alert.common.message.model.LinkableItem;
 
 @Component
 public class DigestMessageContentProcessor extends MessageContentProcessor {
-    private final Map<ItemOperation, BiFunction<Map<CategoryKey, CategoryItem>, CategoryItem, Void>> operationFunctionMap;
+    private final MessageContentCollapser messageContentCollapser;
 
     @Autowired
-    public DigestMessageContentProcessor() {
+    public DigestMessageContentProcessor(final MessageContentCollapser messageContentCollapser) {
         super(FormatType.DIGEST);
-        final BiFunction<Map<CategoryKey, CategoryItem>, CategoryItem, Void> addFunction = createAddFunction();
-        final BiFunction<Map<CategoryKey, CategoryItem>, CategoryItem, Void> deleteFunction = createDeleteFunction();
-        operationFunctionMap = new EnumMap<>(ItemOperation.class);
-        operationFunctionMap.put(ItemOperation.ADD, addFunction);
-        operationFunctionMap.put(ItemOperation.UPDATE, addFunction);
-        operationFunctionMap.put(ItemOperation.DELETE, deleteFunction);
+        this.messageContentCollapser = messageContentCollapser;
     }
 
     @Override
     public List<AggregateMessageContent> process(final List<AggregateMessageContent> messages) {
-        final List<AggregateMessageContent> collapsedTopicList = new ArrayList<>(messages.size());
-        for (final AggregateMessageContent topic : messages) {
-            final Map<CategoryKey, CategoryItem> categoryDataCache = new LinkedHashMap<>();
-            topic.getCategoryItems().forEach(item -> processOperation(categoryDataCache, item));
-
-            final Optional<AggregateMessageContent> collapsedContent = rebuildTopic(topic, categoryDataCache.values());
-            if (collapsedContent.isPresent()) {
-                collapsedTopicList.add(collapsedContent.get());
-            }
+        final List<AggregateMessageContent> collapsedMessages = messageContentCollapser.collapse(messages);
+        final Map<LinkableItem, List<AggregateMessageContent>> topicInfoToTopics = groupByTopic(collapsedMessages);
```

Comment:
Could this be a list of objects? It doesn't look like we use the map functionality at all here and just loop through each item.

Reply:
I think the map is appropriate here